docs/design.md

```diff
@@ -0,0 +1,128 @@
+
+## State transfer
+
+State transfer refers to the mechanism via which out of date replicas can
+retrieve already committed data efficiently and bring themselves in sync with
+the rest of the cluster. State transfer is done outside of the ordering and
+replication protocol performed by SBFT, and can often be achieved by data transfer
+between only two replicas.
+
+As state transfer is outside of the core SMR protocol, it may be implemented in
+any number of ways, depending upon the application using concord-bft.
+Concord-bft treats application operations and data to be replicated/executed
+upon commit as opaque. Since the data being stored by the application is opaque,
+and can be arbitrarily large, concord-bft cannot provide an efficient mechanism
+for transferring such state, and relies on application specific
+implementations for optimal distribution.
+
+For applications to implement state transfer, and have concord-bft properly be
+able to initiate state transfer when needed, and also know when there is more
+state that needs to be transferred, there must be hooks in both the application
+code and concord-bft code that allow interaction between the application and
+bftengine in a generic manner. This is the purpose of the state transfer
+interfaces in
+[IStateTransfer.hpp](../bftengine/include/bftengine/IStateTransfer.hpp).
+
+The `IStateTransfer` interface is implemented by the application developer for a given
+state transfer/checkpoint mechanism. It is called by the main replica thread of
+the core bftengine implementation. The concord-bft library contains two
+implementations of `IStateTransfer`:
+[SimpleStateTransfer](../bftengine/src/simplestatetransfer) and
+[BcStateTransfer](../bftengine/src/bcstatetransfer). These will be described
+further in the detailed design section for state transfer.
+
+The `IReplicaForStateTransfer` interface allows the state transfer
+implementations to call functions in the bftengine replica.
+
+## Metadata Storage
+
+The bftengine has internal metadata that must be persisted to disk. The
+[MetadataStorage](../bftengine/include/bftengine/MetadataStorage.hpp) interface
+should be implemented by the application developer for a given storage system
+such as RocksDB.
+
+## Replica
+
+The [Replica](../bftengine/include/bftengine/Replica.hpp) interface is the API
+used to instantiate a new replica given specific implementations of the other
+public interfaces.
+
+## RequestsHandler
+
+The [RequestsHandler](../bftengine/include/bftengine/Replica.hpp) interface
+contains a single callback function to be implemented by the application and
+called by the bftengine when an operation at a given sequence number has
+committed. This is the hook that allows arbitrary SMR implemented in concord-bft
+to result in execution of application specific operations at all replicas in a
+total order.
+
+# ReplicaConfig
+ReplicaConfig contains all configurable attributes of concord-bft and should be
```

Comment:
Don't we still also have the separate communication configuration, and, if so, should we weaken this statement about "all configurable attributes"?

Reply:
Good point!